R/plot_crf_completeness_data_preparation.R

```diff
@@ -0,0 +1,130 @@
+    url <- "https://clinicaltrials-pre.plymouth.ac.uk/api/"
+  } else{
+    url <- "https://clinicaltrials.plymouth.ac.uk/api/"
+  }
+  
+  #load the instrument data
+  formData <- list("token"=token,
+                   content='instrument',
+                   format='csv',
+                   returnFormat='csv'
+  )
+  response <- httr::POST(url, body = formData, encode = "form")
+  crf_data <- httr::content(response)
+  
+  #load the instrument-event mapping data
+  formData <- list("token"=token,
+                   content='formEventMapping',
+                   format='csv',
+                   'arms[0]'='1',
+                   returnFormat='csv'
+  )
+  response <- httr::POST(url, body = formData, encode = "form")
+  event_instrument_mapping <- httr::content(response)
+  names(event_instrument_mapping)[names(event_instrument_mapping) == "form"] <- "instrument_name"
+  all_event_data <- merge(event_instrument_mapping, crf_data, by = "instrument_name", all.x = TRUE)
+  
+  #add completeness column names to the dataset
+  all_event_data <- all_event_data %>%
+    mutate(completeness_column_name = paste0(instrument_name, "_complete"))
+  
+  #load event data from API
+  formData <- list("token"=token,
+                   content='event',
+                   format='csv',
+                   returnFormat='csv'
+  )
+  response <- httr::POST(url, body = formData, encode = "form")
+  event_data <- httr::content(response)
+  
+  all_event_data <- merge(all_event_data, event_data, by = "unique_event_name", all.x = TRUE)
+  
+  #subset the dataset so that only the record_id, category and required completeness columns are present
+  if(category == "Allocation"){
+    selected_columns <- c("Allocation", "redcap_event_name", all_event_data$completeness_column_name)
+    table_dataset <- dataset %>%
+      select(all_of(selected_columns))  
+  }
+  if(category == "Site"){
+    selected_columns <- c("Site",  "redcap_event_name", all_event_data$completeness_column_name)
+    table_dataset <- dataset %>%
+      select(all_of(selected_columns))  
+  }
+  
+  
+  result_list <- list()
+  for (timepoint_name in timepoint_names) {
+    subset_rows <- all_event_data[all_event_data$unique_event_name == timepoint_name, ]
+    
+    completeness_columns <- subset_rows$completeness_column_name
+  
+    subset_table <- table_dataset[table_dataset$redcap_event_name == timepoint_name, ]
+    completeness_values <- apply(subset_table[, completeness_columns], 1, function(row) {
+      if (any(!is.na(row)) && ((any(row %in% c(1, 2)) && any(row == 0)) || (any(row == 1) && any(row == 2)))) {
+        return("Partially complete")
+      } else if (all(row == 2, na.rm = TRUE) || all(is.na(row) | row == 2)) {
+        return("Complete")
+      } else if (all(is.na(row)) || all(row == 0) || all(is.na(row) | row == 0)) {
+        return("Not started")
+      } else {
+        return("Undefined")  # You can adjust this based on your specific needs
+      }
+    }) 
```

Comment:
Turn this into a function: 

calculate_completeness <- function(row) {
  if (any(!is.na(row)) && ((any(row %in% c(1, 2)) && any(row == 0)) || (any(row == 1) && any(row == 2)))) {
    return("Partially complete")
  } else if (all(row == 2, na.rm = TRUE) || all(is.na(row) | row == 2)) {
    return("Complete")
  } else if (all(is.na(row)) || all(row == 0) || all(is.na(row) | row == 0)) {
    return("Not started")
  } else {
    return("Undefined")
  }
}

then you can use:     completeness_values <- apply(subset_table[, completeness_columns], 1, calculate_completeness)

nesting bespoke 'business logic', inside another packages function (apply) is generally not the best practice. Better to seperate out the function and then reference that directly instead.